Source/ARTHTTPPaginatedResponse.m

```diff
@@ -0,0 +1,139 @@
+//
+//  Created by Ricardo Pereira on 17/08/2018.
+//  Copyright © 2018 Ably. All rights reserved.
+//
+
+#import "ARTHTTPPaginatedResponse.h"
+
+#import "ARTHttp.h"
+#import "ARTAuth.h"
+#import "ARTRest+Private.h"
+#import "ARTPaginatedResult+Private.h"
+#import "ARTNSMutableURLRequest+ARTPaginated.h"
+#import "ARTNSHTTPURLResponse+ARTPaginated.h"
+#import "ARTEncoder.h"
+#import "ARTConstants.h"
+
+@implementation ARTHTTPPaginatedResponse {
+    NSHTTPURLResponse *_response;
+}
+
+- (instancetype)initWithResponse:(NSHTTPURLResponse *)response
+                           items:(NSArray *)items
+                            rest:(ARTRest *)rest
+                        relFirst:(NSMutableURLRequest *)relFirst
+                      relCurrent:(NSMutableURLRequest *)relCurrent
+                         relNext:(NSMutableURLRequest *)relNext
+               responseProcessor:(ARTPaginatedResultResponseProcessor)responseProcessor {
+    self = [super initWithItems:items rest:rest relFirst:relFirst relCurrent:relCurrent relNext:relNext responseProcessor:responseProcessor];
+    if (self) {
+        _response = response;
+    }
+    return self;
+}
+
+- (NSInteger)statusCode {
+    return _response.statusCode;
+}
+
+- (BOOL)success {
+    return _response.statusCode >= 200 && _response.statusCode < 300;
+}
+
+- (NSInteger)errorCode {
+    NSString *code = [_response.allHeaderFields valueForKey:ARTHttpHeaderFieldErrorCodeKey];
+    return [code integerValue];
+}
+
+- (NSString *)errorMessage {
+    NSString *message = [_response.allHeaderFields valueForKey:ARTHttpHeaderFieldErrorMessageKey];
+    return message;
+}
+
+- (NSDictionary<NSString *,NSString *> *)headers {
+    return _response.allHeaderFields;
+}
+
+- (void)first:(void (^)(ARTHTTPPaginatedResponse *_Nullable result, ARTErrorInfo *_Nullable error))callback {
+ART_TRY_OR_REPORT_CRASH_START(self.rest) {
+    if (callback) {
+        void (^userCallback)(ARTHTTPPaginatedResponse *_Nullable result, ARTErrorInfo *_Nullable error) = callback;
+        callback = ^(ARTHTTPPaginatedResponse *_Nullable result, ARTErrorInfo *_Nullable error) {
+            ART_EXITING_ABLY_CODE(self.rest);
+            dispatch_async(self.userQueue, ^{
+                userCallback(result, error);
+            });
+        };
+    }
+
+    [self.class executePaginated:self.rest withRequest:self.relFirst callback:callback];
+} ART_TRY_OR_REPORT_CRASH_END
+}
+
+- (void)next:(void (^)(ARTHTTPPaginatedResponse *_Nullable result, ARTErrorInfo *_Nullable error))callback {
+ART_TRY_OR_REPORT_CRASH_START(self.rest) {
+    if (callback) {
+        void (^userCallback)(ARTHTTPPaginatedResponse *_Nullable result, ARTErrorInfo *_Nullable error) = callback;
+        callback = ^(ARTHTTPPaginatedResponse *_Nullable result, ARTErrorInfo *_Nullable error) {
+            ART_EXITING_ABLY_CODE(self.rest);
+            dispatch_async(self.userQueue, ^{
+                userCallback(result, error);
+            });
+        };
+    }
```

Comment:
This `if` looks the same as implemented in `first:`. Worth extracting it in separate method?